cargo/workspace.bzl

```diff
@@ -0,0 +1,250 @@
+  "repository_name": {name},
+  "packages": [
+      {packages}
+  ],
+  "cargo_toml_files": {{
+      {cargo_toml_files}
+  }},
+  "overrides": {{
+      {overrides}
+  }},
+  "repository_template": "{repository_template}",
+  "target_triples": [
+    {targets}
+  ],
+  "cargo": "{cargo}"
+}}""".format(
+        name = "\"{}\"".format(repository_ctx.attr.name),
+        packages = ",\n".join([artifact for artifact in repository_ctx.attr.packages]),
+        cargo_toml_files = ",\n".join(['"{}": "{}"'.format(ct, repository_ctx.path(ct)) for ct in repository_ctx.attr.cargo_toml_files]),
+        overrides = ",\n".join(["\"{}\": {}".format(oname, ovalue) for (oname, ovalue) in repository_ctx.attr.overrides.items()]),
+        repository_template = repository_ctx.attr.repository_template,
+        targets = ",\n".join(['"{}"'.format(target) for target in repository_ctx.attr.supported_targets]),
+        cargo = repository_ctx.path(cargo_label),
+    )
+
+    input_path = "_{name}.json".format(name = repository_ctx.attr.name)
+
+    # For debugging or working on changes to the resolver, you can set something like:
+    #   export RULES_RUST_RESOLVER_URL_OVERRIDE=file:///path/to/rules_rust/cargo/crate_universe_resolver/target/release/resolver
+    resolver_url = repository_ctx.os.environ.get("RULES_RUST_CRATE_UNIVERSE_RESOLVER_URL_OVERRIDE", None)
+    if resolver_url and resolver_url.startswith("file://"):
+        sha256_result = repository_ctx.execute(["sha256sum", resolver_url[7:]])
+        resolver_sha = sha256_result.stdout[:64]
+
+        resolver_path = repository_ctx.path("resolver")
+        repository_ctx.download(
+            url = resolver_url,
+            sha256 = resolver_sha,
+            output = resolver_path,
+            executable = True,
+        )
+    else:
+        # TODO: Avoid downloading both the linux and darwin one each time.
+        if repository_ctx.os.name == "linux":
+            resolver_path = repository_ctx.path(repository_ctx.attr._resolver_script_linux)
+        elif repository_ctx.os.name == "mac os x":
+            resolver_path = repository_ctx.path(repository_ctx.attr._resolver_script_darwin)
+
+    repository_ctx.file(input_path, content = input_content)
+
+    args = [
+        resolver_path,
+        "--input_path",
+        input_path,
+        "--output_path",
+        repository_ctx.path("defs.bzl"),
+        "--repo-name",
+        repository_ctx.attr.name,
+    ]
+    if lockfile_path != None:
+        args.append("--lockfile")
+        str(args.append(lockfile_path))
+    if repository_ctx.os.environ.get("RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE", "false") == "true":
+        args.append("--update-lockfile")
+
+    path_binaries = [
+        repository_ctx.path(rustc_label),
+    ]
+
+    result = repository_ctx.execute(
+        args,
+        environment = {
+            "RUST_LOG": "info",
+
+            # The resolver invokes `cargo metadata` which relies on `rustc` being on the $PATH
```

Comment:
Does cargo metadata actually invoke rustc? What for?

I worry that having to use host platform to run cargo metadata will somehow spoil the generated repositories and there will be incompatibilities when host platform is not exactly exec platform. Do you use remote execution with rules rust and crate_universe? Were there ever problems with this?

Reply:
> Does cargo metadata actually invoke rustc? What for?

It specifically invokes `rustc -vV` which on my current system outputs:
```console
% rustc -vV
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-apple-darwin
release: 1.49.0
```

I did some tracing through cargo, and it doesn't actually use this output for anything, it just passes it to some methods which conditionally use it in other codepaths. I filed https://github.com/rust-lang/cargo/issues/9295 to try to remove this requirement in cargo.

> I worry that having to use host platform to run cargo metadata will somehow spoil the generated repositories and there will be incompatibilities when host platform is not exactly exec platform. Do you use remote execution with rules rust and crate_universe? Were there ever problems with this?

For what we're using it for, cargo resolves the superset of all platforms (which is why you end up with dependencies on things like `winapi` on all platforms in your `Cargo.lock`), and only narrows down when it gets around to the actual fetching and building. We are indeed using remote execution with `crate_universe` and have not run into any issues :)